### PR TITLE
Je/sms export final

### PR DESCRIPTION
**Pull request by emord**
Replacing #14631 

@benrudolph @gcapalbo 

This export is static, so it doesn't allow the user to create new exports and there's not a list view. Also all domains will use "new" export

cc @proteusvacuum

### CODE REVIEW

File: corehq/apps/export/tests/test_sms_export.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
+import datetime
+import json
+import uuid
+from django.test.testcases import SimpleTestCase
+from mock import patch
+
+from couchexport.models import Format
+
+from corehq.apps.export.export import (
```
Comment (dimagimon) [inline corehq/apps/export/tests/test_sms_export.py]: F401 'ExportFile' imported but unused
F401 '_Writer' imported but unused
F401 '_get_writer' imported but unused
F401 '_write_export_instance' imported but unused

File: corehq/apps/export/views.py [hunk 2]
```diff
@@ -2283,6 +2286,57 @@ def get_filters(self, filter_form_data, mobile_user_and_group_slugs):
...
+    def _get_filter_form(self, filter_form_data):
+        filter_form = self.filter_form_class(
+            self.domain_object, self.timezone, filter_form_data,
+        )
+        if not filter_form.is_valid():
+            raise ExportFormValidationException()
+        return filter_form
+
+    def _get_export(self, domain, export_id):
+        user = self.request.couch_user
```
Comment (dimagimon) [inline corehq/apps/export/views.py]: F841 local variable 'user' is assigned to but never used

File: corehq/apps/export/models/new.py [hunk 3]
```diff
@@ -920,6 +933,53 @@ def get_filters(self):
...
 
 
+class SMSExportInstance(ExportInstance):
+    type = SMS_EXPORT
+    identifier = None
+    name = "Messages"
+
+    @classmethod
+    def _new_from_schema(cls, schema):
+        columns = [
```
Comment (emord) [inline corehq/apps/export/models/new.py]: Could also make these nametuple but this felt readable to me since it's only used here

File: corehq/apps/export/views.py [hunk 4]
```diff
@@ -2283,6 +2286,56 @@ def get_filters(self, filter_form_data, mobile_user_and_group_slugs):
         return form_filters
 
 
+class DownloadNewSmsExportView(GenericDownloadNewExportMixin, BaseDownloadExportView):
+    urlname = 'new_export_download_sms'
+    page_title = ugettext_noop("Export SMS")
+    form_or_case = None  # todo: remove this property from exports
```
Comment (emord) [inline corehq/apps/export/views.py]: I think this should get removed from exports totally and replaced with something like `export_type`, which already exists. Want to dig into this, but was going to separate that out into a different PR

Comment (benrudolph) [inline corehq/apps/export/views.py]: 👍 didn't even know that existed

File: corehq/apps/export/models/new.py (hunk 3 above)
Comment (benrudolph) [inline corehq/apps/export/models/new.py]: why not define all of these in the system_properties.py file?

File: corehq/apps/export/models/new.py [hunk 5]
```diff
@@ -920,6 +933,53 @@ def get_filters(self):
...
+            main_table.columns.append(ExportColumn(
+                label="Message Log ID",
+                item=ExportItem(
+                    path=[PathNode(name='_id')]
+                ),
+                selected=True,
+            ))
+
+        instance = cls(domain=schema.domain, tables=[main_table])
+        cls._insert_system_properties(instance.domain, schema.type, instance.tables[0])
```
Comment (benrudolph) [inline corehq/apps/export/models/new.py]: this should already get called on `generate_instance_from_schema`

File: corehq/apps/export/views.py [hunk 6]
```diff
@@ -2283,6 +2286,56 @@ def get_filters(self, filter_form_data, mobile_user_and_group_slugs):
         return form_filters
 
 
+class DownloadNewSmsExportView(GenericDownloadNewExportMixin, BaseDownloadExportView):
```
Comment (benrudolph) [inline corehq/apps/export/views.py]: i would've expected to have seen more views like the edit view and create view. are those not for this PR?

**Review by benrudolph — COMMENTED** [review]
looking good, couple of questions

File: corehq/apps/export/models/new.py (hunk 3 above)
Comment (emord) [inline corehq/apps/export/models/new.py]: Fair I'll update that. originally didn't think of them as "system" properties, but since it's a static thing, i guess they all are

File: corehq/apps/export/views.py (hunk 4 above)
Comment (emord) [inline corehq/apps/export/views.py]: Going to wait on a refactor if this still exists after old exports are no longer supported

File: corehq/apps/export/views.py (hunk 6 above)
Comment (emord) [inline corehq/apps/export/views.py]: Since this is a "static" report I didn't make those. When you click "SMS Exports" you're automatically taken to the download page because there's nothing to edit or create

File: corehq/apps/export/models/new.py (hunk 5 above)
Comment (emord) [inline corehq/apps/export/models/new.py]: Looks like that doesn't happen because I'm creating the schema without any `group_schemas` since all properties are system properties. Do you think it's worth moving those properties over to a `group_schema` in `SMSExportDataSchema`?

File: corehq/apps/export/models/new.py [hunk 7]
```diff
@@ -920,6 +933,53 @@ def get_filters(self):
...
+    identifier = None
+    name = "Messages"
+
+    @classmethod
+    def _new_from_schema(cls, schema):
+        columns = [
+            ("Contact Type", "couch_recipient_doc_type", DOC_TYPE_TRANSFORM),
+            ("Contact ID", "couch_recipient", None),
+            ("Timestamp", "date", None),
+            ("Owner Name", "couch_recipient", OWNER_ID_TRANSFORM),
```
Comment (gcapalbo) [inline corehq/apps/export/models/new.py]: i don't think the owner_id transform will work here since it doesn't handle cases right? the couch_recipient could point to a case, mobile worker, web user, or be blank

File: corehq/apps/export/esaccessors.py [hunk 8]
```diff
@@ -26,6 +27,12 @@ def get_case_export_base_query(domain, case_type):
             .sort("opened_on"))
 
 
+def get_sms_export_base_query(domain):
+    return (SMSES()
+            .domain(domain)
+            .sort("date"))
```
Comment (gcapalbo) [inline corehq/apps/export/esaccessors.py]: i forgot to mention, we should also exclude any messages that are outbound and had errors. we can do that by excluding those messages with `direction` equal to [OUTGOING](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/sms/models.py) and `processed` False

File: corehq/apps/export/models/new.py [hunk 9]
```diff
@@ -920,6 +933,53 @@ def get_filters(self):
...
+                    selected=True
+                )
+                for col in columns
+            ],
+        )
+        if schema.include_metadata:
+            main_table.columns.append(ExportColumn(
+                label="Message Log ID",
+                item=ExportItem(
+                    path=[PathNode(name='_id')]
```
Comment (gcapalbo) [inline corehq/apps/export/models/new.py]: does this name reference work? i know we call it `couch_id` on the SMS (postgres) model and that should get copied to to ES too

**Comment by gcapalbo** [comment]
@emord once you're done should we put this up on staging for UAT/QA before release?

File: corehq/apps/export/models/new.py (hunk 9 above)
Comment (emord) [inline corehq/apps/export/models/new.py]: `_id` will be the id of the document in ES. I thought those would be the same because they are the same for the couch documents that get moved over.

It looks like `SMS` is fully on SQL now? In that case I think this should be changed to `id` since we would want the postgres id? We could (maybe should) change the indexer to make elasticsearch and postgres the same

File: corehq/apps/export/esaccessors.py (hunk 8 above)
Comment (emord) [inline corehq/apps/export/esaccessors.py]: `processed` isn't included in the sms mapping so this will require a reindex https://github.com/dimagi/commcare-hq/blob/8667a4159ad39d3307f92a3812d74006d610fd1b/corehq/pillows/mappings/sms_mapping.py#L36

File: corehq/apps/export/models/new.py (hunk 9 above)
Comment (gcapalbo) [inline corehq/apps/export/models/new.py]: yeah SMS models are only in postgres now.  i'd prefer to keep couch_id as the uuid for an SMS though because we may need to shard it some day and then we couldn't use the integer id anymore

i don't know if _id in the ES doc will be the right thing, it might be if it gets populated by [this](https://github.com/dimagi/commcare-hq/blob/2e0d3a06d394753b188b3fbcdc97b3aaffbaf901/corehq/pillows/sms.py#L47)

File: corehq/apps/export/esaccessors.py (hunk 8 above)
Comment (gcapalbo) [inline corehq/apps/export/esaccessors.py]: hmm, ok. there's not too many of them anyway, maybe it's easier to just exclude them when building the export?

File: corehq/pillows/mappings/sms_mapping.py [hunk 10]
```diff
@@ -1,65 +1,52 @@
...
-                    "type": "custom",
-                    "tokenizer": "whitespace",
-                    "filter": ["lowercase"]
-                },
-            }
-        }
+SMS_INDEX = es_index("smslogs_2017-02-09")
+SMS_MAPPING = {
+    '_meta': {
+       'comment': 'Autogenerated [corehq.apps.sms.models.SMSLog] mapping from ptop_generate_mapping 09/23/2013',
```
Comment (dimagimon) [inline corehq/pillows/mappings/sms_mapping.py]: E121 continuation line under-indented for hanging indent

File: corehq/apps/export/tests/test_sms_export.py [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+                            u'Direction',
+                            u'Message',
+                            u'Type',
+                        ],
+                        u'rows': rows_value,
+                    }
+                }
+            )
+
+    @patch('corehq.apps.export.transforms._cached_case_id_to_case_name')
```
Comment (dimagimon) [inline corehq/apps/export/tests/test_sms_export.py]: F811 redefinition of unused 'test_export_doc_type_transform' from line 126

File: corehq/apps/export/models/new.py (hunk 9 above)
Comment (emord) [inline corehq/apps/export/models/new.py]: Offlined. _id is the correct thing to use

File: corehq/apps/export/models/new.py (hunk 7 above)
Comment (emord) [inline corehq/apps/export/models/new.py]: Yeah. you're right. I misread the owner id transform. Created a new one https://github.com/dimagi/commcare-hq/pull/14823/commits/6c8b135a70731bfc2c7e650da7767b42cf9b4469#diff-0111f3dfa65bf39568e054504f4430afR23

**Comment by emord** [comment]
This is on staging now

File: corehq/apps/export/models/new.py (hunk 5 above)
Comment (benrudolph) [inline corehq/apps/export/models/new.py]: ah gotcha, understood. i don't think we're in the greatest situation, essentially we're bypassing logic meant to add all of this stuff in smart ways, but since this is actually a static export, it doesn't actually matter that much. begs the question of whether we should model static exports slightly differently. in any case i think this is a fine workaround. thanks for looking into it

File: corehq/apps/export/views.py (hunk 6 above)
Comment (benrudolph) [inline corehq/apps/export/views.py]: right makes sense thanks

File: corehq/apps/export/models/new.py (hunk 5 above)
Comment (emord) [inline corehq/apps/export/models/new.py]: Yeah. I agree that we should look into making a distiniction between "static" exports and form and case ones. I'm hesitant to start that refactor with base classes before we make another static export though.

File: corehq/apps/es/tests/test_sms.py [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+        json_output = {
+            "query": {
+                "filtered": {
+                    "filter": {
+                        "and": [
+                            {"term": {"domain.exact": "demo"}},
+                            {
+                                "not": {
+                                    "and": (
+                                        {"term": {"direction": "o"}},
```
Comment (gcapalbo) [inline corehq/apps/es/tests/test_sms.py]: i noticed this in the filter too, that direction is lowercase o instead of capital O. is that because ES is doing case-insensitive matching?

Comment (emord) [inline corehq/apps/es/tests/test_sms.py]: Right. By default it is case insensitive https://github.com/dimagi/commcare-hq/blob/master/corehq/pillows/base.py#L9

Comment (gcapalbo) [inline corehq/apps/es/tests/test_sms.py]: :+1: thanks

**Comment by gcapalbo** [comment]
looks great @emord, i just ran it on staging. would it be hard to convert the timestamps to the timezone of the project rather than have them be utc?

**Comment by benrudolph** [comment]
there was an effort to do that here: https://github.com/dimagi/commcare-hq/pull/13100. got bogged down in product stuff

**Review by benrudolph — APPROVED** [review]

**Comment by gcapalbo** [comment]
this is ready to go from my side - @dimagi/product / @sheelio just let us know once you have a chance to check it out on staging for uat

**Comment by sheelio** [comment]
Did a pass. Looks generally good - minor feedback below. 
- After clicking "Download" and the file is downloaded, I still see the "Filters Inactive While Download Is In Progress" overlay.  I think form or case exports remove that bubble once done. 
- The data file is inside a folder inside a zip (guessing we can drop the folder if easy)
- Checking this doesn't integrate with the saved / daily export stuff (which is fine)

**Comment by emord** [comment]
@sheelio 

Just tried a form export and the overlay is still present once the download is done

I think the data file should be in a zip because text is highly compressible so there will likely be good gains for a csv file

It doesn't integrate with saved/daily export, but I could look into that if we determine it's important

**Comment by sheelio** [comment]
@emord 

Okay, all sounds good.  I was mainly pushing to get rid of the extra directory that you get inside the CSV (but its really minor so okay to just merge as is).

**Comment by emord** [comment]
kicked off the preindex on prod. Should be ready for merging on monday

**Comment by gcapalbo** [comment]
:+1: great thanks @emord

**Comment by emord** [comment]
preindexing finished. this should be good to merge.

FYI @NoahCarnahan I didn't run this on india/swiss, but those should be quick to index
